core/src/main/java/org/infinispan/factories/NamedExecutorsFactory.java

```diff
@@ -107,13 +108,18 @@ public Object construct(String componentName) {
 
    private ThreadFactory createThreadFactoryWithDefaults(GlobalConfiguration globalCfg, final String componentName,
                                                          ExecutorServiceType type) {
-      if (type.isNonBlocking()) {
-         return new DefaultNonBlockingThreadFactory(null, getDefaultThreadPrio(componentName),
-               DefaultThreadFactory.DEFAULT_PATTERN, globalCfg.transport().nodeName(), shortened(componentName));
+      switch (type) {
+         case BLOCKING:
+            return new BlockingThreadFactory("ISPN-blocking-thread-group", getDefaultThreadPrio(componentName),
```

Comment:
Couldn't we have a single `DefaultThreadFactory` class again, with the thread group as a constructor parameter?

Reply:
I think so, yeah. Originally this was for the thread creation.

Reply:
Actually, it isn't quite that simple as we would lose a check to ensure that the thread factory is non blocking when registering a non blocking executor. I can do an instanceof check and invoke a method instead though.

Reply:
Actually this ends up being much more troublesome as `DefaultThreadFactory` is in core which would have to move down to commons as well. I think leaving it how I had it is probably best unfortunately.

Reply:
> Actually, it isn't quite that simple as we would lose a check to ensure that the thread factory is non blocking when registering a non blocking executor. I can do an instanceof check and invoke a method instead though.

Maybe I'm missing something, but why do we need a check, if the blocking threads and the non-blocking threads are exactly the same, and only the thread group is different.

> Actually this ends up being much more troublesome as DefaultThreadFactory is in core which would have to move down to commons as well.

I'm definitely missing something here, I just see that `NamedExecutorsFactory`, `BlockingThreadFactory`, and `NonBlockingThreadFactory` are also in core.

Reply:
https://github.com/infinispan/infinispan/blob/master/commons/all/src/main/java/org/infinispan/commons/executors/BlockingThreadPoolExecutorFactory.java#L67 is the guy who checks to make sure that when you have an executor that is non blocking that you also have a thread factory that produces non blocking threads.

Reply:
> > Actually, it isn't quite that simple as we would lose a check to ensure that the thread factory is non blocking when registering a non blocking executor. I can do an instanceof check and invoke a method instead though.
> 
> Maybe I'm missing something, but why do we need a check, if the blocking threads and the non-blocking threads are exactly the same, and only the thread group is different.

They aren't exactly the same, we have to make sure all threads of a given executor are either never blocked or can be.

Reply:
I had to try to remove `BlockingThreadFactory` and `NonBlockingThreadFactory` myself to see the problem, but I think I have a solution here:

https://github.com/infinispan/infinispan/compare/master...danberindei:ISPN-11510_thread_type_detection

I'm looking forward to having the programmatic configuration as simple as the XML configuration, with its `<blocking-thread-pool/>` and `<non-blocking-thread-pool/>` elements.

> [...] we have to make sure all threads of a given executor are either never blocked or can be.

Sure, but the check now involves only the thread's group, not the thread itself.

Reply:
> I'm looking forward to having the programmatic configuration as simple as the XML configuration, with its <blocking-thread-pool/> and <non-blocking-thread-pool/> elements.

Oops, I thought we had merged the `<thread-factory/>` element into `<blocking-thread-pool/>` and `<non-blocking-thread-pool/>`. I guess the XML configuration is just as complicated as the programmatic one :)

Reply:
> Oops, I thought we had merged the `<thread-factory/>` element into `<blocking-thread-pool/>` and `<non-blocking-thread-pool/>`. I guess the XML configuration is just as complicated as the programmatic one :)

No, this is something I had thought of doing, but didn't really look into it. I would love if we didn't have it, it is really annoying to say the least.

Reply:
> I had to try to remove `BlockingThreadFactory` and `NonBlockingThreadFactory` myself to see the problem, but I think I have a solution here:
> 
> [master...danberindei:ISPN-11510_thread_type_detection](https://github.com/infinispan/infinispan/compare/master...danberindei:ISPN-11510_thread_type_detection)
> 
> I'm looking forward to having the programmatic configuration as simple as the XML configuration, with its `<blocking-thread-pool/>` and `<non-blocking-thread-pool/>` elements.
> 
> > [...] we have to make sure all threads of a given executor are either never blocked or can be.
> 
> Sure, but the check now involves only the thread's group, not the thread itself.

So the changes here delay the check until runtime and inside the executor itself. This isn't as friendly from an error perspective as what I have at the moment as the error is immediately seen in the thread that is bootstrapping the components, which will bubble up quite well. I am not sure how this other error will propagate if at all.

Reply:
Fair enough, hopefully we'll simplify the config at some point and make that check unnecessary anyway.